sdk/appcenter/src/main/java/com/microsoft/appcenter/SessionContext.java

```diff
@@ -0,0 +1,220 @@
+import java.util.TreeMap;
+import java.util.UUID;
+
+import static com.microsoft.appcenter.AppCenter.LOG_TAG;
+
+/**
+ * Persistent session history.
+ */
+public class SessionContext {
+
+    /**
+     * Key used in storage to persist sessions.
+     */
+    private static final String STORAGE_KEY = "sessions";
+
+    /**
+     * Maximum number of sessions to persist the state.
+     */
+    private static final int STORAGE_MAX_SESSIONS = 10;
+
+    /**
+     * Separator used for persistent storage format.
+     */
+    private static final String STORAGE_KEY_VALUE_SEPARATOR = "/";
+
+    /**
+     * Singleton.
+     */
+    private static SessionContext sInstance;
+
+    /**
+     * Past and current session identifiers sorted by session starting timestamp (ascending).
+     */
+    private final NavigableMap<Long, SessionInfo> mSessions = new TreeMap<>();
+
+    /**
+     * App launch timestamp. We could use the real process start time and not SDK start time.
+     * But there is no Android API to do that it requires executing ps command or reading proc files.
+     * This is used to know minidump files app launch timestamp that are processed after restart.
+     * This is not used for regular managed crashes where that timestamp is maintained by Crashes
+     * and has the same limitation (initialized at SDK start time).
+     */
+    private final long mAppLaunchTimestamp;
+
+    /**
+     * Init.
+     */
+    @WorkerThread
+    private SessionContext() {
+
+        /* Try loading past sessions from storage. */
+        mAppLaunchTimestamp = System.currentTimeMillis();
+        Set<String> storedSessions = StorageHelper.PreferencesStorage.getStringSet(STORAGE_KEY);
+        if (storedSessions != null) {
+            for (String session : storedSessions) {
+                String[] split = session.split(STORAGE_KEY_VALUE_SEPARATOR, -1);
+                try {
+                    long time = Long.parseLong(split[0]);
+                    String rawSid = split[1];
+                    UUID sid = rawSid.isEmpty() ? null : UUID.fromString(rawSid);
+                    long appLaunchTimestamp;
+                    if (split.length > 2) {
+                        appLaunchTimestamp = Long.parseLong(split[2]);
+                    } else {
+
+                        /* Backward compatibility with older SDK storage. Use placeholder. */
+                        appLaunchTimestamp = time;
+                    }
+                    mSessions.put(time, new SessionInfo(time, sid, appLaunchTimestamp));
+                } catch (RuntimeException e) {
+                    AppCenterLog.warn(LOG_TAG, "Ignore invalid session in store: " + session, e);
+                }
+            }
+        }
+        AppCenterLog.debug(LOG_TAG, "Loaded stored sessions: " + mSessions);
+
+        /*
+         * Record a session with no identifier
+         * to avoid correlating log to a session from a previous process.
+         */
+        addSession(null);
```

Comment:
I don't think this is really a big deal, but:

If there are 10 sessions persisted, `SessionContext` will drop the oldest one due to the addition of a `null` session. Is that okay?

Reply:
That's why I increased from 5 to 10.